toggle profile switches instead of individual devices

AaronDavidSchneider opened the pull request:
With this PR, profiles are toggled (by changing the online time from unlimited to never and vice versa). The previous switch for device profiles reassigned devices to groups which has been a huge overkill. 

I tested (as appropriate):
- [ ] Get/Set port switch
- [ ] Get/Set 2.4 Ghz wifi
- [ ] Get/Set 5 Ghz wifi
- [ ] Get/Set guest wifi switch
- [ ] Get/Set call deflections
- [ ] Connectivity sensor
- [x] Yaml Mode

Closes issue #109 

Remaining Question:
- [x] Does a sid (session id) expire? Needs longer term test. Could easily be fixed in underlying library.
Shouldn't be a problem with release 0.2 of fritzprofiles. Could have been one of the problems of the usage of the other library.

- [x] Naming of `CONF_USE_DEVICES` and  `DEFAULT_USE_DEVICES` have been kept untouched. Should they be changed?

AaronDavidSchneider commented:
I updated the error handling to fix issue #70. The problem has been that the error alway occured in __init__ of Fritzboxtools. Therefore we need to grab this error already there.

AaronDavidSchneider commented:
Since the error messages are indeed really helpful, I added an additional error message, when a profile doesn't exist. With this  update, a user can see the following Errors:

Wrong credentials:
![Screenshot 2020-08-26 at 15 39 54](https://user-images.githubusercontent.com/7556347/91310961-cb58a280-e7b2-11ea-8b94-8966a7d13c7f.png)

Attempt to use profile switches, but "Anmeldung im Heimnetzwerk mit Nutzername und Passwort" disabled (fritzconnection doesn't seem to care here, since it uses the fritz-api):
![Screenshot 2020-08-26 at 15 42 06](https://user-images.githubusercontent.com/7556347/91310953-ca277580-e7b2-11ea-94e6-87523f258762.png)

Attempt to use profile switches, but profiles mentioned are not existant:
![Screenshot 2020-08-26 at 15 40 43](https://user-images.githubusercontent.com/7556347/91310959-cac00c00-e7b2-11ea-9def-096b39d9577f.png)

AaronDavidSchneider commented:
Forgot to mention: Latest commit closes #101 and #102.

mammuth commented on custom_components/fritzbox_tools/switch.py:
```diff
@@ -452,44 +433,16 @@ class FritzBoxProfileSwitch(SwitchEntity):
-                "The profile you tried to set does not exist in the fritzbox. "
-                "Please check profile_on and profile_off in your config for errors"
-            )
+        self._is_available = True  # set to False if an error happend during toggling the switch
```
This comment doesn't make sense anymore, right?

AaronDavidSchneider replied:
Yes thats true. I will change that.

mammuth commented on custom_components/fritzbox_tools/switch.py:
```diff
@@ -514,39 +467,24 @@ def available(self) -> bool:
                 self._is_on = False
-            else:
+            elif status == "unlimited":
                 self._is_on = True  # TODO: Decide on default behaviour
```
Shouldn't we also set `_is_available` in `if status == "never"` and `elif status == "unlimited"`? Otherwise, it might be unavailable until the integration gets reloaded, right?

AaronDavidSchneider replied:
I will provide a fix. Thanks for the hint.

mammuth commented on custom_components/fritzbox_tools/switch.py:
```diff
@@ -514,39 +467,24 @@ def available(self) -> bool:
-            )
-        else:
-            await self._async_fetch_update()
+        await self._async_fetch_update()
```
If we don't have any logic in `#async_update` anymore, do we actually still need `#_async_fetch_update` or can't we just move everything here?

AaronDavidSchneider replied:
Thats true. I will update that.

mammuth commented on custom_components/fritzbox_tools/switch.py:
```diff
@@ -573,11 +511,11 @@ async def async_turn_off(self, **kwargs) -> None:
+            state = "unlimited"
         else:
-            state = [[self.device["id1"], self.id_off]]
+            state = "never"
```
For these simple assignments, you could use ternary assignments: `state = "unlimited" if turn_on else "never"`

AaronDavidSchneider replied:
thats a good idea.

mammuth commented on custom_components/fritzbox_tools/translations/de.json:
```diff
@@ -13,12 +13,10 @@
               }
             },
             "setup_devices": {
```
Shouldn't we also change this dictionary key to eg. `setup_profiles`?

AaronDavidSchneider replied:
I will change that.

mammuth commented on custom_components/fritzbox_tools/translations/de.json:
```diff
@@ -13,12 +13,10 @@
-                "title": "Setup FRITZ!Box Tools - device settings",
-                "description": "Device profiles need a list of devices.\nleave it empty if you don't need it.\nDocumentation: https://github.com/mammuth/ha-fritzbox-tools",
+                "title": "Setup FRITZ!Box Tools - profile settings",
+                "description": "Device profiles needs at least one profile.\nleave it empty if you don't need it.\nDocumentation: https://github.com/mammuth/ha-fritzbox-tools",
```
We could probably also translate the messages to German when we already touch the files 🤔

AaronDavidSchneider replied:
Yes... but that sounds like a lot of work.... :D

mammuth reviewed (changes requested):
Looks good overall, nice! I added some smaller comments 😊

AaronDavidSchneider commented on README.md:
```diff
@@ -40,12 +40,10 @@ fritzbox_tools:
-  devices: # Optional. Needed if you want to control the profiles of your network devices.
-    - "Helens-iPhone"
-    - "Aarons-MacBook-Air"
+  profiles: # Optional. Needed if you want to control the profiles of your network devices.
```
Good idea. I will add that

AaronDavidSchneider commented on custom_components/fritzbox_tools/__init__.py:
```diff
@@ -165,24 +153,42 @@ def __init__(
+        
+        try:
+            self.connection = FritzConnection(
+                address=host, port=port, user=username, password=password, timeout=30.0
```
we could extend it to 60 seconds. What do you think?

AaronDavidSchneider commented on custom_components/fritzbox_tools/config_flow.py:
```diff
@@ -160,20 +150,32 @@ async def async_step_setup_options(self, user_input=None):
-            devices = devices.replace(' ', '').split(',')
+        profiles = user_input.get(CONF_PROFILES,DEFAULT_PROFILES)
+        if isinstance(profiles, str):
+            profiles = profiles.replace(' ', '').split(',')
```
probably not. I will test it and try to fix it. Thanks for the hint!

AaronDavidSchneider replied:
Should work now

AaronDavidSchneider commented on custom_components/fritzbox_tools/strings.json:
```diff
@@ -13,12 +13,10 @@
-                    "profile_on": "Device profiles: Profile on",
-                    "profile_off": "Device profiles: Profile off",
-                    "devices": "Device profiles: comma seperated list of devices"
+                    "profiles": "Device profiles: Comma seperated list of Profiles"
```
Good idea!

AaronDavidSchneider commented on custom_components/fritzbox_tools/strings.json:
```diff
@@ -33,7 +31,9 @@
         "error": {
-            "connection_error": "Failed to connect to FRITZ!Box. Check your username and password."
+            "connection_error": "Failed to connect to FRITZ!Box. Check your username and password.",
+            "connection_error_profiles": "Profile switches only work if login via user and password is enabled in the FRITZ!Box",
```
Good idea! I will add it.

AaronDavidSchneider commented:
Thanks for the feedback! I implemented your requests. Also: I changed all names with devices to access profiles. Thats the word used in the english version of the fritzbox (changed my language settings from german to english)
<img width="1157" alt="Screenshot 2020-08-29 at 17 13 52" src="https://user-images.githubusercontent.com/7556347/91640033-06f5a580-ea1b-11ea-9b3a-7193a416bbbf.png">

mammuth commented:
Nice, should I merge and create a pre-release? I think HACS somehow supports this so people who subscribed to beta updates of the integration can get it first.

AaronDavidSchneider commented:
Yes that sounds good!
